llvm/lib/Transforms/Vectorize/VectorCombine.cpp

```diff
@@ -2678,6 +2679,52 @@ bool VectorCombine::shrinkType(llvm::Instruction &I) {
   return true;
 }
 
+/// insert (DstVec, (extract (binop), ExtIdx), InsIdx) -->
+/// shuffl (DstVec, (binop), Mask)
+bool VectorCombine::foldInsExtOfBinOpShuffle(Instruction &I) {
+  Value *DstVec;
+  BinaryOperator *BO;
+  uint64_t ExtIdx, InsIdx;
+  if (!match(&I, m_InsertElt(
+                     m_Value(DstVec),
+                     m_OneUse(m_ExtractElt(m_BinOp(BO), m_ConstantInt(ExtIdx))),
+                     m_ConstantInt(InsIdx))))
+    return false;
+
+  if (!isSafeToSpeculativelyExecute(BO))
+    return false;
+
+  auto *VecTy = cast<FixedVectorType>(I.getType());
+  if (BO->getType() != VecTy)
+    return false;
+
+  unsigned NumElts = VecTy->getNumElements();
+  if (ExtIdx >= NumElts)
+    return false;
+
+  SmallVector<int> Mask(NumElts);
+  std::iota(Mask.begin(), Mask.end(), 0);
```

Comment:
please just use `SmallVector (size_t Size, const T &Value)`

Reply:
Its needs the iota for the ascending shuffle mask - its not a splat.

Reply:
I couldn't catch up what you mean. 
@mshockwave you mean it is better to use SmallVector with initial value? or it can be replace from std::iota?
@RKSimon what is mean splat?

Reply:
A splat is when the same value is assigned to every element - but for this shuffle mask you need `<0, 1, 2, 3, ..., N-1>` which is what iota will give you.

Reply:
I see! thakns for letting me know

Reply:
my bad, I misunderstood what `std::iota` does